src/osd/PrimaryLogPG.cc

```diff
@@ -10022,6 +10085,62 @@ void PrimaryLogPG::issue_repop(RepGather *repop, OpContext *ctx)
   for (auto &&entry: ctx->log) {
     projected_log.add(entry);
   }
+
+  for (map<pg_shard_t, pg_missing_t>::iterator i = peer_missing.begin();
+           i != peer_missing.end();
+           ++i) {
+    dout(10) << __func__ << " shard " << i->first << " before missing " << (i->second).get_items() << dendl;
+  }
+  bool requires_missing_loc = false;
+  for (set<pg_shard_t>::iterator i = async_recovery_targets.begin();
+       i != async_recovery_targets.end();
+       ++i) {
+    if (*i == get_primary() || !peer_missing[*i].is_missing(soid)) continue;
+    requires_missing_loc = true;
+    for (auto &&entry: ctx->log) {
+      peer_missing[*i].add_next_event(entry);
+    }
+  }
+
+  for (map<pg_shard_t, pg_missing_t>::iterator i = peer_missing.begin();
+           i != peer_missing.end();
+           ++i) {
+    dout(10) << __func__ << " shard " << i->first << " after add next missing " << (i->second).get_items() << dendl;
+  }
+
+  for (set<pg_shard_t>::const_iterator i = acting_recovery_backfill.begin();
+       i != acting_recovery_backfill.end();
+       ++i) {
+    pg_shard_t peer(*i);
+    if (peer == pg_whoami) continue;
+    if (async_recovery_targets.count(peer) && peer_missing[peer].is_missing(soid)) {
+      for (auto &&entry: ctx->log) {
+	missing_loc.add_missing(soid, ctx->at_version, eversion_t(), entry.is_delete());
+      }
+    }
+  }
+
+  dout(10) << __func__ << " missing_loc before: " << missing_loc.get_locations(soid) << dendl;
+
+  if (requires_missing_loc) {
+    // clear out missing_loc
+    set<pg_shard_t> peers(missing_loc.get_locations(soid));
+    for (set<pg_shard_t>::iterator r = peers.begin();
+         r != peers.end();
+         ++r) {
+      pg_shard_t peer(*r);
+      missing_loc.remove_location(soid, peer);
```

Comment:
this should only contain peers, so it'd be a little more efficient to add a MissingLoc::clear_location(soid) call that just does clear() on the locations instead of looping